File: client-android/app/src/main/java/stasis/client_android/activities/fragments/sources/SourceRules.kt
package stasis.client_android.activities.fragments.sources

import stasis.client_android.lib.collection.rules.Rule
import stasis.client_android.lib.collection.rules.SourceUri
import stasis.client_android.lib.model.server.datasets.DatasetDefinitionId

object SourceRules {
    fun matching(rules: List<Rule>, scheme: String): List<Rule> =
        rules.filter { SourceUri.scheme(it.source) == scheme }

    fun isEnabled(rules: List<Rule>, scheme: String): Boolean =
        enabledSets(rules, scheme).isNotEmpty()

    fun isInDefault(rules: List<Rule>, scheme: String): Boolean =
        null in enabledSets(rules, scheme)

    fun allSets(rules: List<Rule>): Set<DatasetDefinitionId?> =
        rules.mapTo(mutableSetOf(null)) { it.definition }

    fun enabledSets(rules: List<Rule>, scheme: String): Set<DatasetDefinitionId?> =
        rules.filter { it.operation is Rule.Operation.Include && SourceUri.scheme(it.source) == scheme }
            .map { it.definition }
            .toSet()

    fun setsMissing(rules: List<Rule>, scheme: String): Set<DatasetDefinitionId?> =
        allSets(rules) - enabledSets(rules, scheme)

    fun defaultLibrarySchemes(rules: List<Rule>): Set<String> =
        rules.filter { it.definition == null && it.operation is Rule.Operation.Include }
            .mapNotNull { SourceUri.scheme(it.source) }
            .toSet()
}

File: client-ios/app/StasisClient/Views/Rules/SourceRules.swift
import Foundation
import StasisClientLib

enum SourceRules {
    static func matching(_ rules: [Rule], scheme: String) -> [Rule] {
        rules.filter { SourceUri.scheme($0.source) == scheme }
    }

    static func isEnabled(_ rules: [Rule], scheme: String) -> Bool {
        !enabledSets(rules, scheme: scheme).isEmpty
    }

    static func isInDefault(_ rules: [Rule], scheme: String) -> Bool {
        enabledSets(rules, scheme: scheme).contains(nil)
    }

    static func allSets(_ rules: [Rule]) -> Set<DatasetDefinitionId?> {
        var sets: Set<DatasetDefinitionId?> = [nil]
        for rule in rules { sets.insert(rule.definition) }
        return sets
    }

    static func enabledSets(_ rules: [Rule], scheme: String) -> Set<DatasetDefinitionId?> {
        Set(
            rules
                .filter { $0.operation == .include && SourceUri.scheme($0.source) == scheme }
                .map { $0.definition }
        )
    }

    static func setsMissing(_ rules: [Rule], scheme: String) -> Set<DatasetDefinitionId?> {
        allSets(rules).subtracting(enabledSets(rules, scheme: scheme))
    }

    static func defaultLibrarySchemes(_ rules: [Rule]) -> Set<String> {
        Set(
            rules
                .filter { $0.definition == nil && $0.operation == .include }
                .compactMap { SourceUri.scheme($0.source) }
        )
    }
}
